File: packages/domain/src/consolidation.ts
/**
 * Two half-loads on one truck.
 *
 * A 12-tonne consignment on a 30-tonne trailer pays for the trailer and wastes
 * 18 tonnes of it. The shipper knows they are overpaying and the carrier knows
 * they are underloaded, and neither can fix it alone: matching two strangers'
 * part-loads is exactly the thing a marketplace is for and a phone call is not.
 *
 * `matching.ts` ranks whole loads against a truck. This asks a different
 * question — which *pairs* fit together — and it is stricter, because a
 * consolidation that goes wrong strands two consignments instead of one.
 */

import { distance, type Metres } from './geo.ts';
import { add, percent, type Kobo } from './money.ts';
import { CAPACITY, type TruckClass } from './pricing.ts';
import type { LoadSummary } from './search.ts';

/**
 * How far apart two pickups may be.
 *
 * 50 km. Beyond that the collection detour eats the saving, and the second
 * shipper is waiting on a truck that is somewhere else entirely.
 */
export const PICKUP_SPREAD_M: Metres = 50_000;

/**
 * And two deliveries.
 *
 * 80 km — more generous than the pickup, because by then both consignments are
 * aboard and the detour costs a drive rather than a rendezvous.
 */
export const DROP_SPREAD_M: Metres = 80_000;

/**
 * How much of the truck must actually be used.
 *
 * 70%. Below that the pair is still not worth the coordination: two shippers,
 * two sets of paperwork, two consignees and two chances of a delay, for a
 * trailer that is still mostly air.
 */
export const MINIMUM_FILL = 0.7;

/**
 * What consolidating saves, and who gets it.
 *
 * The pair pays less than two whole-truck fares and more than one. 30% off each
 * — so both shippers see a real reduction — while the carrier still carries
 * 140% of a single fare for one run. Nobody is doing anybody a favour, which is
 * why it works.
 */
export const SHIPPER_DISCOUNT_PCT = 30;

export interface PairLoad extends LoadSummary {
  readonly origin_: { readonly lat: number; readonly lon: number };
  readonly destination_: { readonly lat: number; readonly lon: number };
}

export type PairRefusal =
  | 'too_heavy'
  | 'pickups_too_far'
  | 'drops_too_far'
  | 'wrong_truck'
  | 'too_empty';

export type PairVerdict =
  | {
      readonly ok: true;
      readonly fill: number;
      readonly collectM: Metres;
      readonly deliverM: Metres;
    }
  | { readonly ok: false; readonly reason: PairRefusal; readonly detail: string };

/**
 * Whether two loads can share a truck.
 *
 * Answers with a reason, like every other refusal in this package: a carrier
 * looking at a pair that *nearly* works needs to know which of the five things
 * is wrong, because one of them they might solve with a phone call.
 */
export function canShare(
  a: PairLoad,
  b: PairLoad,
  truck: TruckClass,
): PairVerdict {
  if (a.truckClass !== b.truckClass) {
    return {
      ok: false,
      reason: 'wrong_truck',
      detail: 'These two ask for different kinds of truck.',
    };
  }

  const tonnes = (a.weightKg + b.weightKg) / 1_000;
  const capacity = CAPACITY[truck];

  if (tonnes > capacity) {
    return {
      ok: false,
      reason: 'too_heavy',
      detail: `${Math.round(tonnes)} t together, and this truck takes ${capacity} t.`,
    };
  }

  const fill = tonnes / capacity;
  if (fill < MINIMUM_FILL) {
    return {
      ok: false,
      reason: 'too_empty',
      detail:
        `Together they still only fill ${Math.round(fill * 100)}% of the truck — ` +
        'not worth two sets of paperwork.',
    };
  }

  const collect = distance(
    { ...a.origin_, accuracy: 0, at: a.readyFrom },
    { ...b.origin_, accuracy: 0, at: b.readyFrom },
  );

  if (collect > PICKUP_SPREAD_M) {
    return {
      ok: false,
      reason: 'pickups_too_far',
      detail: `${Math.round(collect / 1_000)} km between the two pickups.`,
    };
  }

  const deliver = distance(
    { ...a.destination_, accuracy: 0, at: a.readyFrom },
    { ...b.destination_, accuracy: 0, at: b.readyFrom },
  );

  if (deliver > DROP_SPREAD_M) {
    return {
      ok: false,
      reason: 'drops_too_far',
      detail: `${Math.round(deliver / 1_000)} km between the two deliveries.`,
    };
  }

  return { ok: true, fill, collectM: collect, deliverM: deliver };
}

export interface Pairing {
  readonly a: PairLoad;
  readonly b: PairLoad;
  readonly fill: number;
  /** What each shipper pays. */
  readonly shipperPays: readonly [Kobo, Kobo];
  /** What the carrier collects for the run. */
  readonly carrierGets: Kobo;
}

export function price(a: PairLoad, b: PairLoad, fill: number): Pairing {
  const discount = (offered: Kobo) => subtractPct(offered, SHIPPER_DISCOUNT_PCT);
  const paysA = discount(a.offered);
  const paysB = discount(b.offered);

  return {
    a,
    b,
    fill,
    shipperPays: [paysA, paysB],
    carrierGets: add(paysA, paysB),
  };
}

function subtractPct(amount: Kobo, pct: number): Kobo {
  return (amount - percent(amount, pct)) as Kobo;
}

/**
 * Every pair worth proposing, fullest first.
 *
 * Quadratic, and deliberately so: a load board a carrier is looking at is a few
 * dozen rows, and every clever index would need maintaining for a saving
 * nobody can perceive.
 */
export function pairs(
  loads: readonly PairLoad[],
  truck: TruckClass,
): readonly Pairing[] {
  const found: Pairing[] = [];

  for (let i = 0; i < loads.length; i++) {
    for (let j = i + 1; j < loads.length; j++) {
      const a = loads[i];
      const b = loads[j];
      if (a === undefined || b === undefined) continue;

      const verdict = canShare(a, b, truck);
      if (!verdict.ok) continue;

      found.push(price(a, b, verdict.fill));
    }
  }

  return found.sort((x, y) => y.fill - x.fill);
}

File: server/src/Backhaul.Domain/Market/Consolidation.cs
using Backhaul.Domain.Money;
using Backhaul.Domain.Pricing;
using Backhaul.Domain.Tracking;

namespace Backhaul.Domain.Market;

/// <summary>A load, as the pairing engine needs it.</summary>
public sealed record PairLoad(
    Guid Id,
    string OriginName,
    string DestinationName,
    string Cargo,
    double WeightKg,
    Kobo Offered,
    DateTimeOffset ReadyFrom,
    TruckClass TruckClass,
    string? ShipperTier,
    double OriginLat,
    double OriginLon,
    double DestinationLat,
    double DestinationLon);

public enum PairRefusal
{
    TooHeavy,
    PickupsTooFar,
    DropsTooFar,
    WrongTruck,
    TooEmpty,
}

public abstract record PairVerdict
{
    public sealed record Ok(double Fill, double CollectM, double DeliverM) : PairVerdict;

    public sealed record No(PairRefusal Reason, string Detail) : PairVerdict;
}

public sealed record Pairing(
    PairLoad A,
    PairLoad B,
    double Fill,
    Kobo PaysA,
    Kobo PaysB,
    Kobo CarrierGets);

/// <summary>
/// Two part-loads, one truck.
/// </summary>
/// <remarks>
/// Mirrors <c>packages/domain/src/consolidation.ts</c>. Every refusal carries
/// a reason for the same reason the rest of this package does: a carrier
/// looking at a pair that nearly works needs to know which of the five things
/// is wrong, because one of them they might solve with a phone call.
/// </remarks>
public static class Consolidation
{
    /// <summary>How far apart two pickups may be.</summary>
    public static readonly double PickupSpreadM = 50_000;

    /// <summary>How far apart two deliveries may be.</summary>
    public static readonly double DropSpreadM = 80_000;

    /// <summary>Below this share of the truck, two sets of paperwork is not worth it.</summary>
    public static readonly double MinimumFill = 0.7;

    /// <summary>What each shipper saves by sharing.</summary>
    public static readonly int ShipperDiscountPct = 30;

    public static PairVerdict CanShare(PairLoad a, PairLoad b, TruckClass truck)
    {
        if (a.TruckClass != b.TruckClass)
        {
            return new PairVerdict.No(PairRefusal.WrongTruck, "These two ask for different kinds of truck.");
        }

        var tonnes = (a.WeightKg + b.WeightKg) / 1_000;
        var capacity = (double)Trucks.Capacity[truck];

        if (tonnes > capacity)
        {
            return new PairVerdict.No(
                PairRefusal.TooHeavy,
                $"{Round(tonnes)} t together, and this truck takes {capacity} t.");
        }

        var fill = tonnes / capacity;
        if (fill < MinimumFill)
        {
            return new PairVerdict.No(
                PairRefusal.TooEmpty,
                $"Together they still only fill {Round(fill * 100)}% of the truck — " +
                "not worth two sets of paperwork.");
        }

        var collect = (double)Geo.Distance(
            new Position(a.OriginLat, a.OriginLon, 0, a.ReadyFrom),
            new Position(b.OriginLat, b.OriginLon, 0, b.ReadyFrom));

        if (collect > PickupSpreadM)
        {
            return new PairVerdict.No(
                PairRefusal.PickupsTooFar,
                $"{Round(collect / 1_000)} km between the two pickups.");
        }

        var deliver = (double)Geo.Distance(
            new Position(a.DestinationLat, a.DestinationLon, 0, a.ReadyFrom),
            new Position(b.DestinationLat, b.DestinationLon, 0, b.ReadyFrom));

        if (deliver > DropSpreadM)
        {
            return new PairVerdict.No(
                PairRefusal.DropsTooFar,
                $"{Round(deliver / 1_000)} km between the two deliveries.");
        }

        return new PairVerdict.Ok(fill, collect, deliver);
    }

    // wired-check: reached only from Pairs() below, which the consolidation
    // parity case exercises, so every figure this produces is compared against
    // TypeScript. Public to match the shape packages/domain exports.
    public static Pairing Price(PairLoad a, PairLoad b, double fill)
    {
        var paysA = a.Offered - a.Offered.Percent(ShipperDiscountPct);
        var paysB = b.Offered - b.Offered.Percent(ShipperDiscountPct);

        return new Pairing(a, b, fill, paysA, paysB, paysA + paysB);
    }

    /// <summary>
    /// Every pair worth proposing, fullest first.
    /// </summary>
    /// <remarks>
    /// Quadratic, and deliberately so: a load board a carrier is looking at is
    /// a few dozen rows, and every clever index would need maintaining for a
    /// saving nobody can perceive.
    /// </remarks>
    public static IReadOnlyList<Pairing> Pairs(IReadOnlyList<PairLoad> loads, TruckClass truck)
    {
        var found = new List<Pairing>();

        for (var i = 0; i < loads.Count; i++)
        {
            for (var j = i + 1; j < loads.Count; j++)
            {
                if (CanShare(loads[i], loads[j], truck) is not PairVerdict.Ok verdict) continue;
                found.Add(Price(loads[i], loads[j], verdict.Fill));
            }
        }

        return found.OrderByDescending(p => p.Fill).ToList();
    }

    /// <summary>JavaScript's <c>Math.round</c>. See <see cref="Matching"/>.</summary>
    private static double Round(double value) => Math.Floor(value + 0.5);
}
